### ISSUE TITLE
Get told I can sell cargo I offloaded at mission depot

### ISSUE THREAD

**Issue by Taomyn** [EDCD/EDDI#771]
## EDDI version in which issue found
3.0.1, under VA

## Steps to reproduce

1. Take "source and return" mission, go pick up necessary cargo and return
2. Drop cargo at depot, not enough to complete, have nothing in hold
3. Travel back to same station to pick up more of the same cargo
4. Dock, wait

## Expected
Nothing other than expected chat from EDDI about the station

## Observed
EDDI tells me I can sell the same cargo even though it's no longer on board - "You can sell your {cargoname} here for a profit........"

I also don't get EDDI telling me I can hand in the cargo at the depot when I arrive with it, but not sure if it's even able to do that, but it does tell me I can sell it.

**Comment by Corysia** [comment]
I hear the same thing.  In addition, when you finish delivering the cargo required for the mission, you're told that the "WING" mission requirements are completed, whether or not it's a WING mission or not.

**Comment by Hoodathunk** [comment]
I'll look at this.

**Comment by Hoodathunk** [comment]
@Corysia , if a mission involves a Cargo Depot, by FDev definition, it as a 'Wing' mission regardless of whether you share it or not. FDev provides no notification as to whether the mission was actually shared and accepted until the first 'Cargo Depot' event with an update type of 'WingUpdate'.

**Comment by Hoodathunk** [comment]
@Taomyn At any point during the execution of the Wing 'Source & Return' mission, was EDDI not running... Did it miss any journal events?

Please provide all journal events associated with the mission... mission accepted, market buy, cargo depot and mission completed, in chronological order.

**Comment by Taomyn** [comment]
@Hoodathunk nope, have it running through VA all the time.

Will try and get logs tonight when I play and upload them for you.

**Comment by Hoodathunk** [comment]
Data provided by @Corysia ...
```{ "timestamp":"2018-08-07T02:40:49Z", "event":"MissionAccepted", "Faction":"Mainani Empire Party", "Name":"Mission_Collect_RankEmp", "LocalisedName":"Source 10 units of Explosives for the Imperial Navy", "Commodity":"$Explosives_Name;", "Commodity_Localised":"Explosives", "Count":10, "DestinationSystem":"Ngalinn", "DestinationStation":"Hickam Survey", "Expiry":"2018-08-08T02:40:25Z", "Wing":false, "Influence":"Med", "Reputation":"Med", "Reward":234156, "MissionID":408693377 }
{ "timestamp":"2018-08-07T03:12:15Z", "event":"CargoDepot", "MissionID":408693377, "UpdateType":"Deliver", "CargoType":"Explosives", "Count":10, "StartMarketID":0, "EndMarketID":3222056192, "ItemsCollected":0, "ItemsDelivered":10, "TotalItemsToDeliver":10, "Progress":0.000000 }
{ "timestamp":"2018-08-07T03:12:27Z", "event":"Promotion", "Empire":2 }
{ "timestamp":"2018-08-07T03:12:27Z", "event":"MissionCompleted", "Faction":"Mainani Empire Party", "Name":"Mission_Collect_RankEmp_name", "MissionID":408693377, "Commodity":"$Explosives_Name;", "Commodity_Localised":"Explosives", "Count":10, "DestinationSystem":"Ngalinn", "DestinationStation":"Hickam Survey", "Reward":234156, "FactionEffects":[ { "Faction":"Mainani Empire Party", "Effects":[ { "Effect":"$MISSIONUTIL_Interaction_Summary_boom_up;", "Effect_Localised":"$#MinorFaction; are experiencing increased growth that could lead to an economic boom", "Trend":"UpGood" } ], "Influence":[ { "SystemAddress":2557753529034, "Trend":"UpGood" } ], "Reputation":"UpGood" } ] }```

**Comment by Taomyn** [comment]
I was able to reproduce the issue with debug logging enabled, I've attached the whole file and you will find here is where I am told I can sell my non-existent cargo:

```
2018-08-07T18:40:38 [Debug] ScriptResolver:resolveScript Turned script {_ Commodity sale check}
{_ Check to see which commodities can be sold here}

{_ Function to find an illegal mission related to the commodity}
{set illegalmission(haulageamounts) to:
    {for haulageamount in haulageamounts:
        {if find(lcase(token(haulageamount.name, '_',2)), "illegal") > -1:
            {return true}
        }
    }
    {return false}
}

{set totalprofit to 0}
{set missioncommodities to []}
{set profitablecommodities to []}
{set stolencommodities to []}
{set unknowncommodities to []}

{for cargo in inventory:
    {if cargo.invariantName != 'Limpet':
        {set marketcommodity to CommodityMarketDetails(cargo.localizedName)}
        {set illegal to illegalmission(cargo.haulageamounts)}
        {set profit to (marketcommodity.sellprice - cargo.price) * cargo.other}

        {if cargo.stolen > 0 && !illegal:
            {set stolencommodities to cat(stolencommodities, [cargo.commodity])}
        |elif cargo.haulage > 0:
            {set missioncommodities to cat(missioncommodities, [cargo.commodity])}
        |elif !marketcommodity && lcase(cargo.category) != "powerplay" && cargo.total > 0:
            {set unknowncommodities to cat(unknowncommodities, [cargo.commodity])}
        |elif marketcommodity && profit > 0:
            {set profitablecommodities to cat(profitablecommodities, [cargo.commodity])}
            {set totalprofit to totalprofit + profit}
        }
    }
}

{_ Let the user know about any commodities they can sell here for a profit}
{if len(profitablecommodities) > 0:
    You can sell your
    {if len(profitablecommodities) = 1:
        {profitablecommodities[0].localizedName}
    |elif len(profitablecommodities) = 2:
        {profitablecommodities[0].localizedName} and {profitablecommodities[1].localizedName}
    |else:
        {set cur to 0}
        {while cur < len(profitablecommodities):
            {if cur = 0:
                {profitablecommodities[cur].localizedName}
            |elif cur < len(profitablecommodities) - 1:
                , {profitablecommodities[cur].localizedName}
            |else:
                , and {profitablecommodities[cur].localizedName}
            }
            {set cur to cur + 1}
        }
    }
    for a profit of {totalprofit} credits.
}

{_ Let the user know about any commodities about which we are unsure if they can sell for a profit}
{if len(unknowncommodities) > 0:
    You could {if len(profitablecommodities) > 0: also } sell your
    {if len(unknowncommodities) = 1:
        {unknowncommodities[0].localizedName}
    |elif len(unknowncommodities) = 2:
        {unknowncommodities[0].localizedName} and {unknowncommodities[1].localizedName}
    |else:
        {set cur to 0}
        {while cur < len(unknowncommodities):
            {if cur = 0:
                {unknowncommodities[cur].localizedName}
            |elif cur < len(profitablecommodities) - 1:
                , {unknowncommodities[cur].localizedName}
            |else:
                , and {unknowncommodities[cur].localizedName}
            }
            {set cur to cur + 1}
        }
    }
    , although check potential profits for {if len(unknowncommodities) > 1: these |else: this} before confirming the transaction.
}

{_ Let the user know about any stolen commodities they can sell on the black market}
{if len(stolencommodities) > 0:
    {if station.hasblackmarket:
        You can fence your stolen
        {if len(stolencommodities) = 1:
            {stolencommodities[0].localizedName}
        |elif len(stolencommodities) = 2:
            {stolencommodities[0].localizedName} and {stolencommodities[1].localizedName}
        |else:
            {set cur to 0}
            {while cur < len(stolencommodities):
                {if cur = 0:
                    {stolencommodities[cur].localizedName}
                |elif cur < len(stolencommodities) - 1:
                    , {stolencommodities[cur].localizedName}
                |else:
                    , and {stolencommodities[cur].localizedName}
                }
                {set cur to cur + 1}
            }
        }
        here.
    |else:
        Reminder: you {OneOf("have stolen cargo on-board", "have stolen goods on-board", "are carrying stolen goods", "are carrying stolen cargo")}.
    }
}

{_ Let the user know about any mission related commodities they can sell on the black market}
{if len(missioncommodities) > 0 && station.hasblackmarket:

    You can {if len(stolencommodities) > 0: also} {OneOf('offload', 'dump')} your mission related
    {if len(missioncommodities) = 1:
        {missioncommodities[0].localizedName}
    |elif len(missioncommodities) = 2:
        {missioncommodities[0].localizedName} and {missioncommodities[1].localizedName}
    |else:
        {set cur to 0}
        {while cur < len(missioncommodities):
            {if cur = 0:
                {missioncommodities[cur].localizedName}
            |elif cur < len(missioncommodities) - 1:
                , {missioncommodities[cur].localizedName}
            |else:
                , and {missioncommodities[cur].localizedName}
            }
            {set cur to cur + 1}
        }
    }
    here.
}
 in to speech You can sell your Cryolite for a profit of 626184 credits. 
```

And here is where I straight afterwards buy 728t of the same thing which is my max capacity:

```
2018-08-07T18:41:21 [Debug] ScriptResolver:resolveScript Turned script {_ Context }
{SetState('eddi_context_last_subject', 'commodity')}
{SetState('eddi_context_last_action', 'purchase')}
{SetState('eddi_context_commodity_name', event.commodity)}
{SetState('eddi_context_commodity_amount', event.amount)}

{if event.amount = 1:
   You have just purchased one tonne of {event.commodity} for {event.price} credits.
|elif event.amount > 1:
    You have just purchased {event.amount} tonnes of {event.commodity} at {event.price} credits per tonne.  Total cost is {Humanise(event.amount * event.price)} credits.
} in to speech You have just purchased 728 tonnes of Cryolite at 1854 credits per tonne. Total cost is on the way to 1 and a half million credits. 
```

Sorry, had to split the log into multiple pastebins due to being so large, and Github would not accept a zipped 2Mb version.

https://pastebin.com/CiXc9uEG
https://pastebin.com/B1RK7trm
https://pastebin.com/ebkEGnsY
https://pastebin.com/9UiWTtNM
https://pastebin.com/VaB39iVG

**Comment by Hoodathunk** [comment]
@Taomyn this is not what I requested, the debug log is not applicable in this instance.

Regardless, if you logout and re-enter the game, the 'Cargo Inventory' event is supposed to repair the Cargo Monitor data, and the Cargo Monitor tab should properly report the cargo in your ship.

If/when this happens again with the mission in question or a future one, please try logging out and back in with your cargo hold empty and verify whether the Cargo Monitor tab shows your hold empty (haulage, stolen, and owned = 0).

Meanwhile, I will revisit the cargo inventory event code.

**Comment by Taomyn** [comment]
Today I was able to dedicate a session to logging for this, so I started up ED and from my session last night I had an empty hold and a partial transport mission to finish. EDDI told me I could sell my non-lethal weapons for a profile. I opened EDDI via VoiceAttack and I took a screen-shot and the figures were a little screwy. Restarting ED made no difference. Proceeded to do pick up more cargo, more missions and things continued to be the same - I even restarted ED again. After clearing my missions and my cargo I did see that EDDI now had me as empty. So began some more missions but for a different commodity, land mines, and the same started to happen.

Anyway I have attached some screen shots and the journal logs from today, hopefully they will help.

![clipboard01](https://user-images.githubusercontent.com/4365929/44147210-6d3b11b4-a093-11e8-868a-77787f8d8abd.png)
![clipboard02](https://user-images.githubusercontent.com/4365929/44147216-75def222-a093-11e8-844b-7175e20bcabb.png)
![clipboard03](https://user-images.githubusercontent.com/4365929/44147217-75f56ba6-a093-11e8-83f5-c7abc90fd71d.png)
![clipboard04](https://user-images.githubusercontent.com/4365929/44147222-7935eb2e-a093-11e8-92b9-4fa46ec908b8.png)
![clipboard05](https://user-images.githubusercontent.com/4365929/44147223-794d08ea-a093-11e8-9a6f-f91a0e9152bb.png)
![clipboard06](https://user-images.githubusercontent.com/4365929/44147224-79630abe-a093-11e8-93f9-043d134b3a06.png)
![clipboard07](https://user-images.githubusercontent.com/4365929/44147229-7e27beb4-a093-11e8-8bf2-3a8a2658e1a9.png)
![clipboard08](https://user-images.githubusercontent.com/4365929/44147230-7e3f278e-a093-11e8-9218-0789b2ec1a88.png)
![clipboard09](https://user-images.githubusercontent.com/4365929/44147231-7e56de06-a093-11e8-9dd2-3c0a977b99a3.png)
![clipboard10](https://user-images.githubusercontent.com/4365929/44147232-7e6b35e0-a093-11e8-9c24-23ab8a48ad0a.png)

https://pastebin.com/Y5THfgJs
https://pastebin.com/bRYUJPLB

**Comment by Hoodathunk** [comment]
Ok, I see a bit about what's going on, but I'm very confused. Why are you using the Cargo Depot to turn in commodities for a simple non-wing 'Collect' mission? Have the game mechanics changed so drastically for a simple source & return mission and I didn't get the memo?

Really scratching my head on this one.

Thanks for the data... I'll parse this out.

**Comment by Taomyn** [comment]
There are wing and non-wing "source and return" missions both use the depot, I've been doing both on my own so "wings" are not a factor. Not sure when FDev brought them in.

**Comment by Hoodathunk** [comment]
This is an undocumented game mechanic change by FDev. Non-wing source and return missions have never required the Cargo Depot... you just go to the origin station with cargo and turn the mission in. The documentation specifically states that the Cargo Depot is for wing missions.

This will take a bit of sorting out.

**Comment by Hoodathunk** [comment]
@Taomyn if it's any consolation, even FDev's Journal event developer wasn't aware of this.

Again, this is going to take some time to sort out.

**Comment by Taomyn** [comment]
Wow, well good luck - let me know if you need more info or to test anything.

**Comment by Hoodathunk** [comment]
Thanks, I will keep that in mind.

I don't do 'source and return missions', so I was a bit blind-sided by this change.

**Comment by Hoodathunk** [comment]
I have a solution that addresses the non-wing 'Collect' (source and return) mission game mechanics change added in 3.1 and the similar change (if you take Howard at his word) coming to non-wing 'Delivery' missions in 3.2

I will PR it in the next day or so. This should be high priority since 3.2 is going live on the 28th of this month.

**Comment by Taomyn** [comment]
Great news, will test it for you as soon as you feel it's ready.

**Comment by Darkcyde13** [comment]
Sorry to open up this again, but it looks like the same problem is occurring with ED v3.2 (EDDI v3.2 under VA 1.7.2).

I've done three cargo delivery missions this evening, one for 132 tonnes of Hydrogen Fuel, one for 135 tonnes of Tea, and 90 tonnes of Tea. EDDI is now telling me I have all of this on board.  My cargo bay is only 96 tonnes of space.  Would you like the Journal entries?

It may be worth noting that Elite is not showing any Wing Delivery Missions at all.  So I'm wondering if they have bugged them with this update?  Of all the stations I've been to tonight, I have only ever seen Assassinate and Massacre missions for Wings, nothing else.

**Comment by Hoodathunk** [comment]
Yes, please include the journal entries. I tried to anticipate them adding solo delivery missions to the mission depot for 3.2, but it's possible I got it wrong. You are using the Mission depot, right?

Also, FYI, it's EDDI 3.0.2, not 3.2.

**Comment by Darkcyde13** [comment]
Yeah, I thought it would be something like that. No worries.  I feel bad for bombarding you with problems already. :/  And yes, I realised my mistake about the EDDI version, sorry about that.

Here's a journal entry for accepting my first mission:
`{ "timestamp":"2018-08-28T19:18:48Z", "event":"MissionAccepted", "Faction":"HR 2283 Independents", "Name":"Mission_Delivery_Boom", "LocalisedName":"Boom time delivery of 132 units of Hydrogen Fuel", "Commodity":"$HydrogenFuel_Name;", "Commodity_Localised":"Hydrogen Fuel", "Count":132, "TargetFaction":"Baltah'Sine Freedom Party", "DestinationSystem":"LTT 2322", "DestinationStation":"Hubble Hub", "Expiry":"2018-08-29T18:44:32Z", "Wing":false, "Influence":"Med", "Reputation":"Med", "Reward":1136647, "MissionID":414461937 }`

Here's the collection event:
`{ "timestamp":"2018-08-28T19:21:22Z", "event":"CargoDepot", "MissionID":414461937, "UpdateType":"Collect", "CargoType":"HydrogenFuel", "CargoType_Localised":"Hydrogen Fuel", "Count":66, "StartMarketID":3223685632, "EndMarketID":3224046592, "ItemsCollected":66, "ItemsDelivered":0, "TotalItemsToDeliver":132, "Progress":0.500000 }`

Here's the delivery event:
`{ "timestamp":"2018-08-28T19:38:31Z", "event":"CargoDepot", "MissionID":414461937, "UpdateType":"Deliver", "CargoType":"HydrogenFuel", "CargoType_Localised":"Hydrogen Fuel", "Count":66, "StartMarketID":3223685632, "EndMarketID":3224046592, "ItemsCollected":66, "ItemsDelivered":66, "TotalItemsToDeliver":132, "Progress":0.000000 }`

The second collect event (I did a bunch of other stuff between these, that's why the timestamp is an hour later):
`{ "timestamp":"2018-08-28T20:37:54Z", "event":"CargoDepot", "MissionID":414461937, "UpdateType":"Collect", "CargoType":"HydrogenFuel", "CargoType_Localised":"Hydrogen Fuel", "Count":66, "StartMarketID":3223685632, "EndMarketID":3224046592, "ItemsCollected":132, "ItemsDelivered":66, "TotalItemsToDeliver":132, "Progress":0.500000 }`

The second deliver event:
`{ "timestamp":"2018-08-28T20:44:50Z", "event":"CargoDepot", "MissionID":414461937, "UpdateType":"Deliver", "CargoType":"HydrogenFuel", "CargoType_Localised":"Hydrogen Fuel", "Count":66, "StartMarketID":3223685632, "EndMarketID":3224046592, "ItemsCollected":132, "ItemsDelivered":132, "TotalItemsToDeliver":132, "Progress":0.000000 }`
Note the deliver event progress is still zero.

And finally the Mission Complete event:
`{ "timestamp":"2018-08-28T20:45:03Z", "event":"MissionCompleted", "Faction":"HR 2283 Independents", "Name":"Mission_Delivery_Boom_name", "MissionID":414461937, "Commodity":"$HydrogenFuel_Name;", "Commodity_Localised":"Hydrogen Fuel", "Count":132, "TargetFaction":"Baltah'Sine Freedom Party", "DestinationSystem":"LTT 2322", "DestinationStation":"Hubble Hub", "Reward":250382, "FactionEffects":[ { "Faction":"Baltah'Sine Freedom Party", "Effects":[ { "Effect":"$MISSIONUTIL_Interaction_Summary_boom_up;", "Effect_Localised":"$#MinorFaction; are experiencing increased growth that could lead to an economic boom", "Trend":"UpGood" } ], "Influence":[ { "SystemAddress":672028042681, "Trend":"UpGood" } ], "Reputation":"UpGood" }, { "Faction":"HR 2283 Independents", "Effects":[ { "Effect":"$MISSIONUTIL_Interaction_Summary_boom_up;", "Effect_Localised":"$#MinorFaction; are experiencing increased growth that could lead to an economic boom", "Trend":"UpGood" } ], "Influence":[ { "SystemAddress":1006926547307, "Trend":"UpGood" } ], "Reputation":"UpGood" } ] }`

If you need anything else let me know. :)

**Comment by Taomyn** [comment]
Damn, I tested the 3.0.2 beta before 3.2 was released and everything looked fine then, so I suppose FDev have once again changed something. That's frustrating.

Once I see a new EDDI beta posted I'll head back to the bubble and help with some testing, if I'm back from vacation by then.

**Comment by Hoodathunk** [comment]
@Taomyn @Darkcyde13 Yes, it's due to the change in game mechanics for solo delivery missions incorporated in 3.2. This should actually be a new issue, but it was pretty straightforward, easily solved, and tested.

Only remaining question is with regards to whether smuggling missions (basically illegal delivery missions) use the mission depot. No smuggling missions in my neck of the galaxy, so help with that determination would be greatly appreciated.

**Comment by Darkcyde13** [comment]
It looks like Smuggling missions are still the same as they were.  At least, the only one I've had the rep to actually take was anyway.  Here's the Missions Accepted:
`{ "timestamp":"2018-08-29T20:51:56Z", "event":"MissionAccepted", "Faction":"Gcirithang Crimson Mafia", "Name":"Mission_Smuggle_Famine", "LocalisedName":"Smuggle 36 units of Narcotics to combat famine", "Commodity":"$BasicNarcotics_Name;", "Commodity_Localised":"Narcotics", "Count":36, "DestinationSystem":"Carcinus", "DestinationStation":"Wye-Delta Station", "Expiry":"2018-08-30T20:55:33Z", "Wing":false, "Influence":"Med", "Reputation":"Med", "Reward":180818, "MissionID":414732731 }`

Here's the Mission Completed:
`{ "timestamp":"2018-08-29T20:59:58Z", "event":"MissionCompleted", "Faction":"Gcirithang Crimson Mafia", "Name":"Mission_Smuggle_Famine_name", "MissionID":414732731, "Commodity":"$BasicNarcotics_Name;", "Commodity_Localised":"Narcotics", "Count":36, "DestinationSystem":"Carcinus", "DestinationStation":"Wye-Delta Station", "Reward":330818, "FactionEffects":[ { "Faction":"Gcirithang Crimson Mafia", "Effects":[ { "Effect":"$MISSIONUTIL_Interaction_Summary_bust_down;", "Effect_Localised":"After showing an economic resurgence, $#MinorFaction; are less likely to experience a bust.", "Trend":"DownGood" } ], "Influence":[ { "SystemAddress":5373540846432, "Trend":"UpGood" } ], "Reputation":"UpGood" }, { "Faction":"Revolutionary Gl 606.1 B Coalition", "Effects":[ { "Effect":"$MISSIONUTIL_Interaction_Summary_bust_up;", "Effect_Localised":"$#MinorFaction; are showing signs of a period of economic bust", "Trend":"UpBad" } ], "Influence":[  ], "Reputation":"None" } ] }`

**Comment by richardbuckle** [comment]
I concur. But it's obvious now that FD consider this something they can change without notice. That's their prerogative, but let's program defensively for it.

**Comment by richardbuckle** [comment]
Oh and we determined that smuggling missions don't (currently) use the depot.
